### PR TITLE
add DDPG algorithm and model

### PR DESCRIPTION
**Pull request by skylian**
- support multiple optimizers
- action perturbation

### CODE REVIEW

File: flare/agent_zoo/simple_rl_agents.py [hunk 1]
```diff
@@ -11,23 +11,30 @@ class SimpleRLAgent(Agent):
...
-                action=actions[0],
-                reward=[r / 100.0 for r in rewards]))
+                sensor=observations[0], action=actions[0], reward=rewards))
 
     def _cts_predict(self, observations, states):
         ## each action is already 2D
         assert len(observations) == 1
         actions, _ = self.predict('RL', inputs=dict(sensor=observations))
-        return [actions.values()[0][0]], []
+        a = actions.values()[0][0]
```
Comment (yu239) [inline flare/agent_zoo/simple_rl_agents.py]: Do not put this in the agent's code. You should really change the algorithm code.

File: flare/framework/computation_task.py [hunk 2]
```diff
@@ -183,15 +188,15 @@ def outermost_step(*args):
                                            init_states=states.values() + next_states.values(),
                                            step_func=outermost_step)
                 costs = dict(zip(self.cost_keys, costs))
+                ### backward and step
+                total_cost, total = sum_cost(costs["cost"])
```
Comment (yu239) [inline flare/framework/computation_task.py]: Keep this file unchanged. Right now the recurrent problem is still unsolved.

File: flare/framework/computation_task.py [hunk 3]
```diff
@@ -147,8 +151,9 @@ def sum_cost(costs):
             return sum(costs), sum(ns)
 
         for i in range(self.alg.iterations_per_batch):
-            self.optim.zero_grad()
             if states:  ## if states is not empty, we apply a recurrent_group first
+                ##TODO: move optimization into algorithm for model with RNN
+                self.optim.zero_grad()
```
Comment (yu239) [inline flare/framework/computation_task.py]: This line should be outside of 'if'

File: flare/framework/tests/test_simple_games.py [hunk 4]
```diff
@@ -48,9 +48,9 @@ def test_gym_games(self):
 
         games = ["MountainCar-v0", "CartPole-v0", "Pendulum-v0"]
         final_rewards_thresholds = [
-            -1.5,  ## drive to the right top in 150 steps (timeout is -2.0)
-            1.5,  ## hold the pole for at least 150 steps
-            -3.0  ## can swing the stick to the top most of the times
+            -150,  ## drive to the right top in 150 steps (timeout is -2.0)
```
Comment (yu239) [inline flare/framework/tests/test_simple_games.py]: Why the changes?

Comment (skylian) [reply flare/framework/tests/test_simple_games.py]: I removed the reward scaling in `Agent`

File: flare/agent_zoo/simple_rl_agents.py (hunk 1 above)
Comment (skylian) [reply flare/agent_zoo/simple_rl_agents.py]: One of the noise used in DDPG is generated by a stochastic process, e.g. `OUNoise`. It is better for each agent to keep its own stochastic process.

File: flare/framework/tests/test_simple_games.py (hunk 4 above)
Comment (yu239) [reply flare/framework/tests/test_simple_games.py]: This file doesn't use Agent. It's a standalone test file.

File: flare/agent_zoo/simple_rl_agents.py (hunk 1 above)
Comment (yu239) [reply flare/agent_zoo/simple_rl_agents.py]: I don't think the agent should "own" the stochastic process. The reason why we want to have algorithm and model is to separate the computation from the general agent control flow.

File: flare/agent_zoo/simple_rl_agents.py [hunk 5]
```diff
@@ -11,23 +11,30 @@ class SimpleRLAgent(Agent):
...
 
     def _cts_store_data(self, observations, actions, states, rewards):
         assert len(observations) == 1 and len(actions) == 1
         self._store_data(
             'RL',
             dict(
-                sensor=observations[0],
-                action=actions[0],
-                reward=[r / 100.0 for r in rewards]))
+                sensor=observations[0], action=actions[0], reward=rewards))
```
Comment (yu239) [inline flare/agent_zoo/simple_rl_agents.py]: If you really want to generalize the reward shaping, you should write an additional API to add into the base class Agent, instead of just directly changing the scaling to 1. I left the / 100.0 there as a temporary solution.

File: flare/algorithm_zoo/ddpg.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    nn.init.uniform_(m.weight, -bound, bound)
+    nn.init.uniform_(m.bias, -bound, bound)
+
+
+def uniform_init(m, bound):
+    nn.init.uniform_(m.weight, -bound, bound)
+    nn.init.uniform_(m.bias, -bound, bound)
+
+
+class ContinuousDeterministic(Model):
```
Comment (yu239) [inline flare/algorithm_zoo/ddpg.py]: Better class name. It's not easy to tell that this class is a Model class just from its name right now.

File: flare/algorithm_zoo/ddpg.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    Model used by DDPG.
+    """
+
+    def __init__(self, input_dims, action_dims):
+        super(ContinuousDeterministic, self).__init__()
+        self.input_dims = input_dims
+        self.action_dims = action_dims
+
+        self.policy_fc1 = nn.Linear(input_dims, 400)
```
Comment (yu239) [inline flare/algorithm_zoo/ddpg.py]: Probably consider generalizing the hidden dims of the layers by passing them as arguments of the class.

File: flare/algorithm_zoo/ddpg.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    def policy_optimizer(self, optim_specs):
+        return optim_specs[0]([{
+            'params': self.policy_fc1.parameters()
+        }, {
+            'params': self.policy_fc2.parameters()
+        }, {
+            'params': self.policy_fc3.parameters()
+        }], **optim_specs[1])
+
+    def critic_optimizer(self, optim_specs):
```
Comment (yu239) [inline flare/algorithm_zoo/ddpg.py]: Comments. Why having this weird behavior?

File: flare/common/distributions.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        return self._param
+
+    def sample(self):
+        return self._param
+
+    def log_prob(self, a):
+        return torch.zeros(a.shape)
+
+
+class GaussianNoise(object):
```
Comment (yu239) [inline flare/common/distributions.py]: The noises are not technically distributions. They should be put in other places.

File: flare/agent_zoo/simple_rl_agents.py [hunk 10]
```diff
@@ -69,3 +64,35 @@ def _cts_predict(self, observations, states):
         actions, next_states = self.predict(
             'RL', inputs=dict(sensor=observations), states=dict(state=states))
         return [actions.values()[0][0]], [next_states.values()[0][0]]
+
+
+class ActionNoiseAgent(Agent):
+    """
+    This class extends `SimpleRLAgent` by applying action noise after
```
Comment (Haichao-Zhang) [inline flare/agent_zoo/simple_rl_agents.py]: Since ActionNoiseAgent is an extension of SimpleRLAgent, why not derive it from SimpleRLAgent?

File: flare/examples/ddpg_example.py [hunk 11]
```diff
@@ -57,7 +57,7 @@
     #    Agent's behavior depends on the actual algorithm being used. Since we 
     #    are using SimpleAC, a proper type of Agent is SimpleRLAgent.
```
Comment (Haichao-Zhang) [inline flare/examples/ddpg_example.py]: SimpleAC, SimpleRLAgent -> might need to be updated? Also appear in later part.

File: flare/agent_zoo/simple_rl_agents.py [hunk 12]
```diff
@@ -69,3 +64,35 @@ def _cts_predict(self, observations, states):
...
+    def __init__(self, env, num_games, action_noise):
+        super(ActionNoiseAgent, self).__init__(env, num_games)
+        self.action_noise = action_noise
+
+    def _cts_store_data(self, observations, actions, states, rewards):
+        assert len(observations) == 1 and len(actions) == 1
+        self._store_data(
+            'RL',
+            dict(
+                sensor=observations[0], action=actions[0], reward=rewards))
```
Comment (Haichao-Zhang) [inline flare/agent_zoo/simple_rl_agents.py]: We could also break line 85 into 3 lines as in simple_rl_agents.py line22-24. This helps to highlight the diff if we add some changes to one type of data, e.g., adding scaling factor to rewards.

File: flare/common/distributions.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        self.sigma = sigma
+
+    def reset(self):
+        pass
+
+    def noise(self):
+        return np.random.normal(size=self.action_dims) * self.sigma + self.mu
+
+
+class OUNoise(object):
```
Comment (Haichao-Zhang) [inline flare/common/distributions.py]: Might be useful to provide the full name and/or a reference in the comment for OUNoise?

File: flare/algorithm_zoo/ddpg.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        }, {
+            'params': self.critic_fc2.parameters()
+        }, {
+            'params': self.critic_fc3.parameters()
+        }], **optim_specs[1])
+
+
+class DDPG(Algorithm):
+    """
+    DDPM impelmentation.
```
Comment (Haichao-Zhang) [inline flare/algorithm_zoo/ddpg.py]: DDPM -> DDPG?

File: flare/agent_zoo/simple_rl_agents.py (hunk 12 above)
Comment (skylian) [reply flare/agent_zoo/simple_rl_agents.py]: the automatic style checks did this.

File: flare/examples/ac_example.py [hunk 15]
```diff
@@ -12,7 +12,7 @@
     """
     game = "CartPole-v0"
 
-    num_agents = 16
+    num_agents = 1
```
Comment (yu239) [inline flare/examples/ac_example.py]: Change this number back to 16

File: flare/algorithm_zoo/ddpg.py (hunk 8 above)
Comment (yu239) [reply flare/algorithm_zoo/ddpg.py]: Not addressed? Why functions policy_optimizer() and critic_optimizer() are implemented in a weird way.

File: flare/agent_zoo/simple_rl_agents.py (hunk 12 above)
Comment (yu239) [reply flare/agent_zoo/simple_rl_agents.py]: You don't have to override this function. Please git pull the most up-to-date code.

File: flare/agent_zoo/simple_rl_agents.py [hunk 16]
```diff
@@ -62,3 +64,35 @@ def _cts_predict(self, observations, states):
...
+
+    def _cts_predict(self, observations, states):
+        ## each action is already 2D
```
Comment (yu239) [inline flare/agent_zoo/simple_rl_agents.py]: Comment confusing. Do you want to say that "the return actions is already a 2D tensor? "

File: flare/agent_zoo/simple_rl_agents.py [hunk 17]
```diff
@@ -62,3 +64,35 @@ def _cts_predict(self, observations, states):
         actions, next_states = self.predict(
             'RL', inputs=dict(sensor=observations), states=dict(state=states))
         return [actions.values()[0][0]], [next_states.values()[0][0]]
+
+
+class ActionNoiseAgent(SimpleRLAgent):
+    """
+    This class extends `SimpleRLAgent` by applying action noise after
```
Comment (yu239) [inline flare/agent_zoo/simple_rl_agents.py]: Add the reason why we generate noise in Agent instead of in Algorithm. Something like "Because each agent needs to maintain its own stochastic process for generating the noise ..."

File: flare/agent_zoo/simple_rl_agents.py [hunk 18]
```diff
@@ -62,3 +64,35 @@ def _cts_predict(self, observations, states):
...
+    def _cts_predict(self, observations, states):
+        ## each action is already 2D
+        assert len(observations) == 1
+        actions, _ = self.predict('RL', inputs=dict(sensor=observations))
+        a = actions.values()[0][0]
+        a = a + self.action_noise.noise()
+
+        return [a], []
+
+    def _reset_env(self):
```
Comment (yu239) [inline flare/agent_zoo/simple_rl_agents.py]: Where is this function called?

Comment (skylian) [reply flare/agent_zoo/simple_rl_agents.py]: This is an overridden to `Agent`'s `_reset_env()`.

File: flare/algorithm_zoo/simple_rnn_algorithms.py [hunk 19]
```diff
@@ -75,21 +75,32 @@ def learn(self, inputs, next_inputs, states, next_states, next_alive,
               actions, next_actions, rewards):
         self.optim.zero_grad()
         assert states
+        # `lens` and `keys` are required by the step function
```
Comment (yu239) [inline flare/algorithm_zoo/simple_rnn_algorithms.py]: Remove this file?

File: flare/algorithm_zoo/simple_algorithms.py [hunk 20]
```diff
@@ -225,7 +242,7 @@ def learn(self, inputs, next_inputs, states, next_states, next_alive,
                                            self.grad_clip)
         self.optim.step()
 
-        return dict(cost=cost), states_update, next_states_update
+        return dict(cost=cost)
```
Comment (yu239) [inline flare/algorithm_zoo/simple_algorithms.py]: You might want to add the condition of states for all the algorithms implemented.

File: flare/framework/tests/test_computation_task.py [hunk 21]
```diff
@@ -89,17 +89,15 @@ def test(input, ct, max):
             width=84, height=84, num_actions=num_actions))
 
         q = SimpleQ(model=SimpleModelQ(
-            dims=dims,
+            dims=[dims],
             num_actions=num_actions,
-            mlp=nn.Sequential(
+            perception_net=nn.Sequential(
```
Comment (yu239) [inline flare/framework/tests/test_computation_task.py]: Why removing the last layer?

Comment (skylian) [reply flare/framework/tests/test_computation_task.py]: the last layer is moved to policy_net or value_net inside SimpleModelQ

File: flare/algorithm_zoo/simple_algorithms.py (hunk 20 above)
Comment (skylian) [reply flare/algorithm_zoo/simple_algorithms.py]: I can use a separate pr to do that
